jcl/src/java.base/share/classes/java/lang/ref/Reference.java

```diff
@@ -196,24 +194,31 @@ public boolean isEnqueued () {
 	@NotCheckpointSafe
 	/*[ENDIF] CRIU_SUPPORT */
 	boolean enqueueImpl() {
-		final ReferenceQueue tempQueue;
 		boolean result;
 		T tempReferent = referent;
-		synchronized(this) {
-			/* Static order for the following code (DO NOT CHANGE) */
-			tempQueue = queue;
-			queue = null;
-			if (state == STATE_ENQUEUED || tempQueue == null) {
-				return false;
-			}
-			result = tempQueue.enqueue(this);
-			if (result) {
-				state = STATE_ENQUEUED;
-				if (null != tempReferent) {
-					reprocess();
+		ReferenceQueue tempQueue = queue;
+		while (true) {
+			synchronized(tempQueue) {
+				synchronized(this) {
+					if (tempQueue != queue) {
```

Comment:
Missing spaces before `(`.
This can't work if `tempQueue` is null (see line 208).

Reply:
Ya, there is also some code in ReferenceQueue.enqueue() that wasn't in the synchronized block. I'll think on it more.